### PR TITLE
generate Makefile.ghc by running `ghc -M` individually for every executable

### PR DESCRIPTION
**Pull request by thielema**
This is a work-around for https://gitlab.haskell.org/ghc/ghc/-/issues/25113.

It should fix https://github.com/ganeti/ganeti/issues/1755.

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -1348,9 +1348,18 @@ HS_MAKEFILE_GHC_SRCS += $(HS_TEST_PROGS:%=%.hs)
...
-	$(GHC) -M -dep-makefile $@ $(DEP_SUFFIXES) $(HFLAGS) $(HFLAGS_DYNAMIC) \
-		-itest/hs \
-		$(HEXTRA_COMBINED) $(HS_MAKEFILE_GHC_SRCS)
+# Run dependency generation for every executable.
+# Up to GHC-9.2 we could perform this in one run of 'ghc -M'.
+# Due to https://gitlab.haskell.org/ghc/ghc/-/issues/25113
+# we have to process every executable manually now.
+# This way, many dependencies in Makefile.ghc are duplicate now
+# but 'make' should cope with this.
+	-mv $@ $@.bak
```
Comment (saschalucas) [inline Makefile.am:1357]: for what purpose is the `.bak` file here?

File: Makefile.am [hunk 2]
```diff
@@ -1348,9 +1348,18 @@ HS_MAKEFILE_GHC_SRCS += $(HS_TEST_PROGS:%=%.hs)
...
+	for prog_hs in $(HEXTRA_COMBINED) $(HS_MAKEFILE_GHC_SRCS); do \
+		$(GHC) -M -dep-makefile $@.part $(DEP_SUFFIXES) $(HFLAGS) $(HFLAGS_DYNAMIC) \
+			-itest/hs $$prog_hs ; \
+		cat $@.part >>$@ ; \
```
Comment (saschalucas) [inline Makefile.am:1361]: Would it hurt to replace `cat` by `sort -u` as you suggested in #1755?

And is `>>` the automake version of normal shell output redirection `>`, or is this a shell append? May be related to why `mv` is used above, if this a append?

**Comment by thielema** [comment]
On Sat, 27 Jul 2024, saschalucas wrote:

> for what purpose is the .bak file here?

I simulate what `ghc -M` does. It does create a .bak file as a first step. 
You also already find Makefile.ghc.bak in the list of temporary files.

**Comment by thielema** [comment]
On Sat, 27 Jul 2024, saschalucas wrote:

> In Makefile.am:
> 
> +	for prog_hs in $(HEXTRA_COMBINED) $(HS_MAKEFILE_GHC_SRCS); do \
> +		$(GHC) -M -dep-makefile ***@***.*** $(DEP_SUFFIXES) $(HFLAGS) $(HFLAGS_DYNAMIC) \
> +			-itest/hs $$prog_hs ; \
> +		cat ***@***.*** >>$@ ; \
> 
> Would it hurt to replace cat by sort -u as you suggested in #1755?

Sorting individual dependency lists makes no difference. I would have to 
sort all dependency blocks together.

> And is >> the automake version of normal shell output redirection >, or is this a shell append?

It is plain shell append. I append the outputs of every run of `ghc -M`.

If I manage to write the dependencies to stdout I could also put the 
for-loop in parentheses and pipe everything in one go to Makefile.ghc, and 
I could also apply `sort --unique` as a last step.

**Comment by saschalucas** [comment]
> I simulate what `ghc -M` does. It does create a .bak file as a first step. You also already find Makefile.ghc.bak in the list of temporary files.

Thanks for clarification. This explains why there is also `Makefile.ghc.part.bak` now. I'm not sure if we need to add this also to temporary files? 

> Sorting individual dependency lists makes no difference.

Of course, my fault not realizing that `cat` is inside the loop.

> If I manage to write the dependencies to stdout I could also put the for-loop in parentheses and pipe everything in one go to Makefile.ghc, and I could also apply `sort --unique` as a last step.

That would be a very elegant way. However don't investigate to much time in that. It's more important to have a workaround as it is.

**Comment by thielema** [comment]
On Sat, 27 Jul 2024, saschalucas wrote:

> Thanks for clarification. This explains why there is also 
> Makefile.ghc.part.bak now. I'm not sure if we need to add this also to 
> temporary files?

Right. Either making it a temporary file or, I think better, deleting it 
after appending.

**Review by saschalucas — APPROVED** [review]
Thanks @thielema.

**Pull request merged**